File: src/CsharpClient/QuixStreams.State/Storage/RocksDbStorage.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using RocksDbSharp;

namespace QuixStreams.State.Storage
{
    /// <summary>
    /// Key/Value storage using RocksDB that implements <see cref="IStateStorage"/> interface.
    /// </summary>
    public class RocksDbStorage : IStateStorage
    {
        /// <summary>
        /// Holds references of RocksDb instances to ensure only one exists and avoid IO lock issue.
        /// Key is the db directory and value is a tuple of RocksDb instance and reference count.
        /// </summary>
        private static readonly Dictionary<string, RocksDbInstance> RocksDbInstances = new ();

        private readonly RocksDb db;
        private readonly string dbDirectory;
        private readonly string keyPrefix;
        private readonly string storageName;

        private WriteBatch writeBatch = new();

        private const char Separator = '/';
        
        /// <summary>
        /// Instantiates a new instance of <see cref="RocksDbStorage"/>
        /// </summary>
        /// <param name="dbDirectory">The directory to open the database</param>
        /// <param name="streamId">Stream id of the storage</param>
        /// <param name="stateName">Stream name of the storage</param>
        public static RocksDbStorage GetStateStorage(string dbDirectory, string streamId, string stateName)
        {
            if (string.IsNullOrEmpty(dbDirectory) || string.IsNullOrEmpty(streamId) || string.IsNullOrEmpty(stateName))
            {
                throw new ArgumentException($"{nameof(dbDirectory)}, {nameof(streamId)} and {nameof(stateName)} cannot be null or empty.");
            }
            
            var storageName = $"{streamId}{Separator}{stateName}";
            return new RocksDbStorage(dbDirectory, storageName);
        }
        
        /// <summary>
        /// Deletes all the states of a stream
        /// </summary>
        /// <param name="dbDirectory">The directory to open the database</param>
        /// <param name="streamId">Stream id of the storage</param>
        public static void DeleteStreamStates(string dbDirectory, string streamId)
        {
            if (string.IsNullOrEmpty(dbDirectory) || string.IsNullOrEmpty(streamId))
            {
                throw new ArgumentException($"{nameof(dbDirectory)} and {nameof(streamId)} cannot be null or empty.");
            }
            
            using var streamStorage = new RocksDbStorage(dbDirectory, storageName: streamId);
            
            streamStorage.Clear();
        }
        
        /// <summary>
        /// Instantiates a new instance of <see cref="RocksDbStorage"/>
        /// </summary>
        /// <param name="dbDirectory">The directory to open the database</param>
        /// <param name="storageName">Name of the storage. Used as a prefix to separate data of other states when they use the same db</param>
        private RocksDbStorage(string dbDirectory, string storageName)
        {
            if (string.IsNullOrEmpty(dbDirectory) || string.IsNullOrEmpty(storageName))
            {
                throw new ArgumentException($"{nameof(dbDirectory)} and {nameof(storageName)} cannot be null or empty.");
            }
            
            this.db = GetOrCreateRocksDbInstance(dbDirectory);
            
            this.dbDirectory = dbDirectory;
            this.keyPrefix = $"{storageName}{Separator}";
            this.storageName = storageName;
        }
        
        /// <inheritdoc/>
        public Task SaveRaw(string key, byte[] data)
        {
            var byteKey = Encoding.UTF8.GetBytes($"{keyPrefix}{key}");
            writeBatch.Put(byteKey, data);
            
            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        public Task<byte[]> LoadRaw(string key)
        {
            var byteKey = Encoding.UTF8.GetBytes($"{keyPrefix}{key}");
            return Task.FromResult(db.Get(byteKey));
        }

        /// <inheritdoc/>
        public Task RemoveAsync(string key)
        {
            var byteKey = Encoding.UTF8.GetBytes($"{keyPrefix}{key}");
            db.Remove(byteKey);
            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        public Task<bool> ContainsKeyAsync(string key)
        {
            var keyWithPrefix = $"{keyPrefix}{key}";
            var contains = db.HasKey(keyWithPrefix);
            return Task.FromResult(contains);
        }

        /// <inheritdoc/>
        public async Task<string[]> GetAllKeysAsync()
        {
            try
            {
                return await Task.Run(() =>
                {
                    var keys = new List<string>();
                    using (var iterator = db.NewIterator())
                    {
                        iterator.Seek(this.keyPrefix);
                        while (iterator.Valid() && iterator.StringKey().StartsWith(this.keyPrefix))
                        {
                            keys.Add(iterator.StringKey().Substring(this.keyPrefix.Length));
                            iterator.Next();
                        }
                    }
                    return keys.ToArray();
                });
            }
            catch (Exception ex)
            {
                throw new Exception("Failed to retrieve all keys", ex);
            }
        }
        
        /// <inheritdoc/>
        public Task ClearAsync()
        { 
            try
            {
                DeleteKeysWithPrefix(this.keyPrefix);
            }
            catch (Exception ex)
            {
                throw new Exception("Failed to clear the storage", ex);
            }
            return Task.CompletedTask;
        }

        private void DeleteKeysWithPrefix(string prefix, string cf = null)
        {
            var z = GetAllKeysAsync().GetAwaiter().GetResult();
            var cfHandle = string.IsNullOrEmpty(cf) ? db.GetDefaultColumnFamily() : db.GetColumnFamily(cf);
            var startKey = Encoding.UTF8.GetBytes(prefix);
            
            using var batch = new WriteBatch();
            using (var iterator = db.NewIterator())
            {
                iterator.Seek(prefix);
                while (iterator.Valid() && iterator.StringKey().StartsWith(prefix))
                {
                    batch.Delete(iterator.Key(), (ulong) iterator.Key().Length, cf: cfHandle);
                    iterator.Next();
                }
            }
            db.Write(batch);

            var x = GetAllKeysAsync().GetAwaiter().GetResult();
        }
        
        /// <inheritdoc/>
        public async Task<int> Count()
        {
            return (await GetAllKeysAsync()).Length;
        }

        /// <inheritdoc/>
        public bool IsCaseSensitive => true;
        
        /// <inheritdoc/>
        public bool CanPerformTransactions => true;
        
        /// <inheritdoc/>
        public void Flush()
        {
            try
            {
                db.Write(writeBatch);
            }
            finally
            {
                writeBatch = new WriteBatch(); //As per documentation, creating a new object is recommended
            }
        }
        
        /// <summary>
        /// Disposes the underlying RocksDB instance
        /// </summary>
        public void Dispose()
        {
            this.writeBatch.Dispose();
            
            if (!RocksDbInstances.TryGetValue(this.dbDirectory, out var dbWithRefCount))
            {
                return;
            }
            
            dbWithRefCount.DecrementRefCount();

            if (dbWithRefCount.IsUsed())
                return;
            
            dbWithRefCount.Db.Dispose(); // closes and disposes the db
            RocksDbInstances.Remove(dbDirectory);
        }
        
        private static RocksDb GetOrCreateRocksDbInstance(string dbDirectory)
        {
            if (RocksDbInstances.TryGetValue(dbDirectory, out var rocksDbInstance))
            {
                rocksDbInstance.IncrementRefCount();
                return rocksDbInstance.Db;
            }
            
            if (!Directory.Exists(dbDirectory))
            {
                Directory.CreateDirectory(dbDirectory);
            }
            
            var dbOptions = new DbOptions().SetCreateIfMissing().SetCreateMissingColumnFamilies();
            var columnFamilies = new ColumnFamilies
            {
                // {stateStoragesCF, new ColumnFamilyOptions()}
            };
            
            var db = RocksDb.Open(dbOptions, dbDirectory, columnFamilies);
            RocksDbInstances.Add(dbDirectory, new RocksDbInstance(db, 1));

            return db;
        }
        
        private class RocksDbInstance
        {
            public RocksDb Db { get; }

            private int refCount;

            public RocksDbInstance(RocksDb db, int initialRefCount)
            {
                Db = db;
                refCount = initialRefCount;
            }

            public void IncrementRefCount()
            {
                Interlocked.Increment(ref refCount);
            }
            
            public void DecrementRefCount()
            {
                var result = Interlocked.Decrement(ref refCount);
                if (result < 0)
                {
                    throw new InvalidOperationException("De-referenced the RocksDB instance more times than tracked");
                }
            }

            public bool IsUsed()
            {
                return this.refCount != 0;
            }
        }
    }
}

File: src/PythonClient/src/quixstreams/statestorages/rocksdbstorage.py
import threading
from typing import Dict, Optional, List, Tuple, Any
from rocksdict import Rdict, WriteBatch
import os


class RocksDbStorage:
    rocks_db_instances: Dict[str, 'RocksDbInstance'] = {}
    RESERVED_CF_NAME = "default"
    SUB_STORAGE_SEPARATOR = "/"

    def __init__(self, db_directory: str, storage_name: str = "default-storage"):
        if not db_directory or not storage_name:
            raise ValueError("db_directory and storage_name cannot be null or empty.")

        if storage_name == self.RESERVED_CF_NAME:
            raise ValueError(f"{storage_name} cannot be '{self.RESERVED_CF_NAME}' as it is reserved by rocksdb.")

        self.db = self.get_or_create_rocks_db_instance(db_directory)

        self.column_family = self.db.get_column_family(storage_name)
        self.db_directory = db_directory
        self.storage_name = storage_name

        self.write_batch = WriteBatch()
        self.use_write_batch = False
        self.sub_storages: Dict[str, RocksDbStorage] = {}

    def save_raw(self, key: str, data: bytes) -> None:
        if self.use_write_batch:
            self.write_batch.put(key, data, self.column_family)
        else:
            self.column_family[key] = data

    def load_raw(self, key: str) -> Optional[bytes]:
        return self.column_family.get(key, None)

    def remove(self, key: str) -> None:
        self.column_family.delete(key)

    def contains_key(self, key: str) -> bool:
        return self.column_family.get(key, None) is not None

    def get_all_keys(self) -> List[str]:
        return list(self.column_family.keys())

    def clear(self) -> None:
        self.db.drop_column_family(self.storage_name)
        self.column_family = self.db.create_column_family(self.storage_name)

    async def count(self) -> int:
        return len(self.get_all_keys())

    @property
    def is_case_sensitive(self) -> bool:
        """Returns whether the storage is case-sensitive"""
        return True

    @property
    def can_perform_transactions(self) -> bool:
        """Return True if transactions are supported, False otherwise."""
        return True

    @staticmethod
    def get_or_create_rocks_db_instance(db_directory: str) -> Rdict:
        if db_directory in RocksDbStorage.rocks_db_instances:
            db, ref_count = RocksDbStorage.rocks_db_instances[db_directory]
            RocksDbStorage.rocks_db_instances[db_directory] = (db, ref_count + 1)
            return db

        if not os.path.exists(db_directory):
            os.makedirs(db_directory)

        db = Rdict(db_directory)
        RocksDbStorage.rocks_db_instances[db_directory] = (db, 1)
        return db

    def start_transaction(self) -> None:
        self.use_write_batch = True

    def commit_transaction(self) -> None:
        for key, value in self.write_batch.items():
            self.db[key] = value
        self.write_batch.clear()
        self.use_write_batch = False

    def __del__(self) -> None:
        if self.db_directory in RocksDbStorage.rocks_db_instances:
            db, ref_count = RocksDbStorage.rocks_db_instances[self.db_directory]
            if ref_count <= 1:
                del RocksDbStorage.rocks_db_instances[self.db_directory]
                del db  # This should close and release the RocksDB instance
            else:
                RocksDbStorage.rocks_db_instances[self.db_directory] = (db, ref_count - 1)

    def get_sub_storage(self, sub_storage_name: str) -> "RocksDbStorage":
        if sub_storage_name in self.sub_storages:
            return self.sub_storages[sub_storage_name]

        sub_storage = RocksDbStorage(self.db_directory, f"{self.storage_name}-{sub_storage_name}")
        self.sub_storages[sub_storage_name] = sub_storage
        return sub_storage

    def remove_sub_storage(self, sub_storage_name: str) -> None:
        if sub_storage_name in self.sub_storages:
            del self.sub_storages[sub_storage_name]

    def dispose(self) -> None:
        # Explicitly release RocksDB resources.
        # The __del__ destructor will take care of decrementing the reference count.
        self.db.close()

    def remove_by_prefix(self, prefix: str) -> None:
        keys_to_remove = [key for key in self.db.keys() if key.startswith(prefix)]
        for key in keys_to_remove:
            del self.db[key]

    def get_by_prefix(self, prefix: str) -> Dict[str, bytes]:
        return {key: self.db[key] for key in self.db.keys() if key.startswith(prefix)}


class RocksDbInstance:
    def __init__(self, db, initial_ref_count):
        self._db = db
        self._ref_count = initial_ref_count
        self._ref_count_lock = threading.Lock()

    @property
    def db(self):
        return self._db

    def increment_ref_count(self):
        with self._ref_count_lock:
            self._ref_count += 1

    def decrement_ref_count(self):
        with self._ref_count_lock:
            self._ref_count -= 1
            if self._ref_count < 0:
                raise Exception("De-referenced the RocksDB instance more times than tracked")

    def is_used(self):
        return self._ref_count != 0